src/java.base/share/classes/jdk/internal/foreign/abi/s390x/sysv/SysVS390xVaList.java

```diff
@@ -33,123 +33,563 @@
+				+ ", __regSaveArea=" + regSaveAreaOfVaList
+				+ '}';
+	}
+
+	static SysVS390xVaList.Builder builder(MemorySession session) {
+		return new SysVS390xVaList.Builder(session);
+	}
+
+	public static non-sealed class Builder implements VaList.Builder {
+		private final MemorySession session;
+		private final List<SimpleVaArg> gprArgs = new ArrayList<>();
+		private final List<SimpleVaArg> fprArgs = new ArrayList<>();
+		private final List<SimpleVaArg> overflowArgs = new ArrayList<>();
+
+		public Builder(MemorySession session) {
+			MemorySessionImpl.toSessionImpl(session).checkValidState();
+			this.session = session;
+		}
+
+		@Override
+		public Builder addVarg(ValueLayout.OfInt layout, int value) {
+			return setArg(layout, value);
+		}
+
+		@Override
+		public Builder addVarg(ValueLayout.OfLong layout, long value) {
+			return setArg(layout, value);
+		}
+
+		@Override
+		public Builder addVarg(ValueLayout.OfDouble layout, double value) {
+			return setArg(layout, value);
+		}
+
+		@Override
+		public Builder addVarg(ValueLayout.OfAddress layout, Addressable value) {
+			return setArg(layout, value.address());
+		}
+
+		@Override
+		public Builder addVarg(GroupLayout layout, MemorySegment value) {
+			return setArg(layout, value);
+		}
+
+		private Builder setArg(MemoryLayout layout, Object value) {
+			Objects.requireNonNull(layout);
+			Objects.requireNonNull(value);
+
+			if (isRegOverflow(TypeClass.isFloatingType(layout), gprArgs.size() + 1, fprArgs.size() + 1)) {
+				overflowArgs.add(new SimpleVaArg(layout, value));
+			} else {
+				TypeClass typeClass = TypeClass.classifyLayout(layout);
+				switch (typeClass) {
+					case INTEGER, POINTER, STRUCT -> {
+						gprArgs.add(new SimpleVaArg(layout, value));
+					}
+					case FLOAT, STRUCT_ONE_FLOAT -> {
+						fprArgs.add(new SimpleVaArg(layout, value));
+					}
+					default -> throw new IllegalStateException("Unsupported TypeClass: " + typeClass);
+				}
+			}
+
+			return this;
+		}
+
+		private boolean isEmpty() {
+			return gprArgs.isEmpty() && fprArgs.isEmpty() && overflowArgs.isEmpty();
+		}
+
+		private void storeArgToMemoryArea(List<SimpleVaArg> vaListArgs, MemorySegment argAreaSegment, boolean isOverflowArea) {
+			MemorySegment argAreaCursor = argAreaSegment;
+
+			for (SimpleVaArg arg : vaListArgs) {
+				MemoryLayout layout = arg.layout;
+				Object argValue = arg.value;
+				TypeClass typeClass = TypeClass.classifyLayout(layout);
+				VarHandle argHandle = TypeClass.classifyVarHandle(layout);
+
+				switch (typeClass) {
+					case STRUCT_ONE_FLOAT, STRUCT -> {
+						MemorySegment struArgValue = (MemorySegment)argValue;
+						/* Set the layout size for the requested struct when the size of the allocated segment
+						 * for struct is greater than the layout size.
+						 */
+						long struArgSize = (struArgValue.byteSize() > layout.byteSize()) ?
+													layout.byteSize() : struArgValue.byteSize();
+
+						switch (typeClass) {
```

Comment:
`typeClass` can only be `STRUCT_ONE_FLOAT` or `STRUCT`, I think you can use `if` and `else`.

Reply:
Updated as suggested above.